chromadb/utils/data_loaders.py

```diff
@@ -22,3 +23,123 @@ def _load_image(self, uri: Optional[URI]) -> Optional[Image]:
+            return list(executor.map(func, iterable))
+        
+# --------------------------- Examples of loaders ---------------------------
+
+FileContents = Union[str, bytes]  # a type for the contents of a file
+
+def load_text(filepath: str) -> str:
+    with open(filepath, "r") as f:
+        return f.read()
+
+def load_bytes(filepath: str) -> bytes:
+    with open(filepath, "rb") as f:
+        return f.read()
+    
+def url_to_contents(
+        url: str, content_extractor=attrgetter('text'), *, params=None, **kwargs
+    ) -> FileContents:
+    import requests
+
+    response = requests.get(url, params=params, **kwargs)
+    response.raise_for_status()
+    return content_extractor(response)
+
+def pdf_file_text(
+        filepath: str, *, page_break_delim='---------------------------'
+    ) -> str:
+    from pypdf import PdfReader  # pip install pypdf
+
+    return page_break_delim.join(
+        page.extract_text() for page in PdfReader(filepath).pages
+    )
+
+# --------------------------- FileLoader ---------------------------
+
+class FileLoader(DataLoader[List[Optional[FileContents]]]):
+    """A DataLoader that loads a list of text files from a list of URIs.
+
+    By default, it loads text files from local files, given URIs that are full file paths.
+    You can specify a prefix thought (usually used to specify a root directory),
+    or a suffix (usually used to specify a file extension).
+
+    Further, you can specify a different `loader`, e.g. to load files from a remote URL, 
+    or to load binary files, or to load text from pdf files, or from S3, or a database, etc.
+
+    Example:
+
+    >>> rootdir = chromadb.__path__[0] + '/'
+    >>> file_loader_1 = FileLoader(prefix=rootdir)
+    >>> file_contents_1 = file_loader_1(['__init__.py', 'types.py'])
+    >>> len(file_contents_1)
+    2
+    >>> 'Embeddings' in file_contents_1[0]  # i.e. __init__.py contains the word 'Embeddings'
+    >>> 'from typing import' in file_contents_1[1]  # i.e. types.py contains the phrase 'from typing import'
+    
+    """
+    
+    def __init__(
+        self,
+        loader=load_text,
+        *,
+        prefix: str = "",
+        suffix: str = "",
+        max_workers: int = multiprocessing.cpu_count(),
+    ) -> None:
+        """
+        Args:
+            loader: A function that takes a (single) URI and returns its contents
+            prefix: A string to prepend to the URI
+            suffix: A string to append to the URI
+            max_workers: The maximum number of threads to use when loading the URIs
+        """
+        self._loader = loader
+        self._prefix = prefix
+        self._suffix = suffix
+        self._max_workers = max_workers
+
+    def _load_file(self, uri: Optional[URI]) -> Optional[FileContents]:
+        if uri is None:
+            return None
+        return self._loader(f"{self._prefix}{uri}{self._suffix}")
+
+    def __call__(self, uris: Sequence[Optional[URI]]) -> List[Optional[FileContents]]:
```

Comment:
Does Optional make sense here? Is there a use case where one can call a file loader without providing any data source? Also, wouldn't pass None to `vectorize` raise an exception? I feel more tests are needed on this.

Reply:
_Does Optional make sense here?_

That's a question for you, or perhaps @atroyn?

I just took [ImageLoader](https://github.com/chroma-core/chroma/blob/caa10f66a4727030deb6d9d644bee70e98d77867/chromadb/utils/data_loaders.py#L9) as the template of how you guys wanted your `DataLoaders` to look like. 

I **did** have the question myself though, and rationalized it as being a design choice on how to handle missing values. Note that in a `Collection.get` for example, if you ask for `uris`, and there is none, you'll get a `[None, ..., None]` list of `None` of the size of the `IDs` array. So I thought (but haven't tested) that you could probably sometimes have'em, and some times not (e.g. `[uri1, None, None, uri2, None]`, thus the `Sequence[Optional[...]]` and `List[Optional[...]]` -- that again, were just emulated (with rationale) from the existing `ImageLoader`.

Reply:
_Also, wouldn't pass None to vectorize raise an exception?_

Yes, if you do `vectorize(func, None)`, but not if you do `vectorize(func, [None, ..., None])`, with a `func` that handles `None`, but that's the responsibility of the user to bring a loader that handles Nones if they're going to have them, no?

Reply:
P.S. The fact though, that `chromadb`'s design allows for arrays for missing values might warrant a tool to transform a function into a `None`-robust function. Something like this decorator:

```python
def _return_none_if_none_input(func):
    """Wraps a method function, making it return `None` if the input is `None`.

    (More precisely, it will return `None` if the first (non-instance) input is `None`.

    >>> class Foo:
    ...     @_return_none_if_none_input
    ...     def bar(self, x, y=1):
    ...         return x + y
    >>> foo = Foo()
    >>> foo.bar(2)
    3
    >>> assert foo.bar(None) is None
    >>> assert foo.bar(x=None) is None

    Note: On the other hand, this will not return `None`, but should:
    ``foo.bar(y=3, x=None)``. To achieve this, we'd need to look into the signature,
    which seems like overkill and I might not want that systematic overhead in my
    methods.
    """

    @wraps(func)
    def _func(self, *args, **kwargs):
        if args and args[0] is None:
            return None
        elif kwargs and next(iter(kwargs.values())) is None:
            return None
        else:
            return func(self, *args, **kwargs)

    return _func
```

This is reminiscent (but is not equivalent) to the [Maybe monad](https://en.wikibooks.org/wiki/Haskell/Understanding_monads/Maybe).